Fix GHC 7.0 build

RyanGlScott opened the pull request:
Added the necessary CPP pragmas to make `semigroupoids` build and test on GHC 7.0.

bergmark commented:
I have restricted the existing versions: https://hackage.haskell.org/package/semigroupoids-5.0.0.1/revisions/

RyanGlScott commented:
Hm. Unfortunately, it occurred to me that this doesn't work on GHC 7.2 either due to `Safe` issues, and to make things worse, the Travis CI builds for 7.0 and 7.2 seem to be borked. It probably won't do much good to fix this unless the success can be duplicated.

If you don't mind, I'm going to try using this pull request to test some possible workarounds for Travis issues. I _think_ the issues are possible to fix with some creativity. If I'm successful, I'll document it here and apply it to my myriad of other compatibility-related pull requests for your projects (namely, for [`tagged`](https://github.com/ekmett/tagged/pull/25), [`comonad`](https://github.com/ekmett/comonad/pull/21), and [`bifunctors`](https://github.com/ekmett/bifunctors/pull/21)).

RyanGlScott commented:
OK, I figured out what is going on with the Travis builds. There are several issues that arise:

* GHC 7.0.1 (and only 7.0.1) complains that `The 'type' field is required for test suites. The available test types are: exitcode-stdio-1.0`. This is a [very old bug](https://mail.haskell.org/pipermail/haskell-cafe/2011-October/096261.html) triggered by `Build_doctests` not being registered as a module in the test suite. However, `cabal sdist` breaks if you add `build_doctests` to the other modules field, so I can't think of a workaround. I removed the GHC 7.0.1 test for now (but GHC 7.0.4 does seem to work, thankfully).
* Some packages would fail to install with the error message `unrecognized option '--disable-benchmarks'`. This is [a regression](https://github.com/haskell/cabal/issues/1004) in `Cabal-1.16`, so I require `Cabal` ≥ 1.18 for all old GHC builds.
* Sometimes, a build will fail with an error message like `The package 'semigroupoids' requires Cabal library version ≥ 1.12 && >=1.10 but no suitable version is installed.`. This is due to [an open `Cabal` bug](https://github.com/haskell/cabal/issues/1666) that pertains to `Custom` setups. A workaround is to manually install `Cabal` before everything else.

I've updated my pull requests for `semigroupoids` and `comonad` your packages with these hacks.
